File: scripts/healer_canary.py
#!/usr/bin/env python3
"""Healer canary rollout — promote/demote patterns, show status.

CLI:
    python3 scripts/healer_canary.py [--status] [--promote PATTERN_ID] [--demote PATTERN_ID] [--json]
"""
from __future__ import annotations

import argparse
import json
import sys
from datetime import datetime, timedelta, timezone
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))

HEALER_LOG = Path.home() / ".claude" / "code-review" / "healer.jsonl"
CIRCUIT_PATH = Path.home() / ".claude" / "code-review" / "healer-circuits.json"
CONFIG_PATH = Path.home() / ".claude" / "code-review" / "config.json"
PATTERNS_PATH = Path(__file__).parent / "healer_patterns.json"


# ---------------------------------------------------------------------------
# Data loading
# ---------------------------------------------------------------------------

def _load_patterns() -> list[dict]:
    try:
        return json.loads(PATTERNS_PATH.read_text())
    except Exception:
        return []


def _read_circuits() -> dict:
    try:
        return json.loads(CIRCUIT_PATH.read_text())
    except Exception:
        return {}


def _load_config() -> dict:
    try:
        return json.loads(CONFIG_PATH.read_text())
    except Exception:
        return {}


def _write_config(config: dict) -> None:
    CONFIG_PATH.parent.mkdir(parents=True, exist_ok=True)
    CONFIG_PATH.write_text(json.dumps(config, indent=2))


def _load_log_entries() -> list[dict]:
    try:
        entries = []
        for line in HEALER_LOG.read_text().splitlines():
            try:
                entries.append(json.loads(line))
            except Exception:
                pass
        return entries
    except Exception:
        return []


# ---------------------------------------------------------------------------
# Stats
# ---------------------------------------------------------------------------

def _compute_stats(pattern_id: str, entries: list[dict], circuits: dict) -> dict:
    pattern_entries = [e for e in entries if e.get("pattern_id") == pattern_id]
    total = len(pattern_entries)
    successful_suggests = sum(1 for e in pattern_entries if e.get("status") == "suggested")
    applied = sum(1 for e in pattern_entries if e.get("status") == "applied")

    cutoff = datetime.now(timezone.utc) - timedelta(days=7)
    aborts_last_7d = 0
    for e in pattern_entries:
        if e.get("status") == "aborted":
            try:
                ts = datetime.fromisoformat(e["timestamp"].replace("Z", "+00:00"))
                if ts >= cutoff:
                    aborts_last_7d += 1
            except Exception:
                pass

    success_count = successful_suggests + applied
    success_rate = success_count / total if total > 0 else 0.0

    circuit_state = circuits.get(pattern_id, {})
    tripped_at = circuit_state.get("tripped_at")
    ever_tripped = circuit_state.get("ever_tripped", False)

    circuit_open = False
    if tripped_at:
        try:
            trip_time = datetime.fromisoformat(tripped_at.replace("Z", "+00:00"))
            if datetime.now(timezone.utc) - trip_time < timedelta(hours=24):
                circuit_open = True
        except Exception:
            pass

    return {
        "total_attempts": total,
        "successful_suggests": successful_suggests,
        "applied": applied,
        "aborts_last_7d": aborts_last_7d,
        "success_rate": round(success_rate, 2),
        "consecutive_failures": circuit_state.get("consecutive_failures", 0),
        "circuit_open": circuit_open,
        "ever_tripped": ever_tripped,
        "tripped_at": tripped_at,
    }


def _check_promotion_criteria(pattern: dict, stats: dict) -> list[str]:
    """Return list of unmet criteria. Empty list = eligible for promotion."""
    reasons = []
    if stats["successful_suggests"] < 5:
        reasons.append(
            f"requires >= 5 successful suggests (have {stats['successful_suggests']})"
        )
    if stats["aborts_last_7d"] > 0:
        reasons.append(
            f"has {stats['aborts_last_7d']} guard failure(s) in the last 7 days"
        )
    if stats["ever_tripped"]:
        reasons.append("circuit has been tripped")
    if pattern.get("requires_confirmation", False):
        reasons.append("requires_confirmation is true")
    return reasons


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _log(entry: dict) -> None:
    try:
        HEALER_LOG.parent.mkdir(parents=True, exist_ok=True)
        with HEALER_LOG.open("a") as f:
            f.write(json.dumps(entry) + "\n")
    except Exception:
        pass


def _emit_alert(level: str, message: str) -> None:
    try:
        from alert_delivery import emit_alert
        emit_alert(level, "healer_canary", message)
    except Exception:
        pass


def _now() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


# ---------------------------------------------------------------------------
# Commands
# ---------------------------------------------------------------------------

def cmd_status(args: argparse.Namespace) -> None:
    patterns = _load_patterns()
    circuits = _read_circuits()
    config = _load_config()
    auto_patterns = config.get("self_heal", {}).get("auto_patterns", [])
    entries = _load_log_entries()

    output = []
    for p in patterns:
        pid = p["id"]
        stats = _compute_stats(pid, entries, circuits)
        mode = "auto" if pid in auto_patterns else "suggest"
        circuit = "open" if stats["circuit_open"] else "closed"
        unmet = _check_promotion_criteria(p, stats)

        output.append({
            "id": pid,
            "mode": mode,
            "circuit": circuit,
            "consecutive_failures": stats["consecutive_failures"],
            "successful_suggests": stats["successful_suggests"],
            "total_attempts": stats["total_attempts"],
            "success_rate": stats["success_rate"],
            "eligible_for_promotion": mode == "suggest" and not unmet,
            "promotion_blockers": unmet if mode == "suggest" else [],
        })

    if args.json:
        print(json.dumps({"patterns": output}))
        return

    for row in output:
        circuit_flag = "OPEN  " if row["circuit"] == "open" else "closed"
        eligible = " [eligible]" if row["eligible_for_promotion"] else ""
        suggests_str = f"{row['successful_suggests']}/5 suggests"
        print(
            f"  {row['id']:30s}  mode={row['mode']:7s}  circuit={circuit_flag}"
            f"  {suggests_str}  rate={row['success_rate']:.0%}{eligible}"
        )


def cmd_promote(args: argparse.Namespace) -> None:
    pattern_id = args.promote
    patterns = _load_patterns()
    pattern = next((p for p in patterns if p["id"] == pattern_id), None)

    if pattern is None:
        print(f"Error: pattern '{pattern_id}' not found", file=sys.stderr)
        sys.exit(1)

    circuits = _read_circuits()
    entries = _load_log_entries()
    stats = _compute_stats(pattern_id, entries, circuits)

    unmet = _check_promotion_criteria(pattern, stats)
    if unmet:
        print(f"Cannot promote '{pattern_id}' — criteria not met:", file=sys.stderr)
        for reason in unmet:
            print(f"  - {reason}", file=sys.stderr)
        sys.exit(1)

    config = _load_config()
    if "self_heal" not in config:
        config["self_heal"] = {}
    auto_patterns = config["self_heal"].get("auto_patterns", [])

    if pattern_id in auto_patterns:
        msg = f"Pattern '{pattern_id}' is already in auto_patterns"
        if args.json:
            print(json.dumps({"promoted": pattern_id, "auto_patterns": auto_patterns, "note": msg}))
        else:
            print(msg)
        return

    auto_patterns.append(pattern_id)
    config["self_heal"]["auto_patterns"] = auto_patterns
    _write_config(config)

    _log({
        "timestamp": _now(),
        "event": "canary_promoted",
        "pattern_id": pattern_id,
        "mode": "auto",
    })
    _emit_alert("info", f"Pattern {pattern_id} promoted to auto-mode")

    if args.json:
        print(json.dumps({"promoted": pattern_id, "auto_patterns": auto_patterns}))
    else:
        print(f"Promoted '{pattern_id}' to auto-mode.")


def cmd_demote(args: argparse.Namespace) -> None:
    pattern_id = args.demote

    config = _load_config()
    if "self_heal" not in config:
        config["self_heal"] = {}
    auto_patterns = config["self_heal"].get("auto_patterns", [])

    if pattern_id not in auto_patterns:
        msg = f"Pattern '{pattern_id}' is not in auto_patterns"
        if args.json:
            print(json.dumps({"demoted": pattern_id, "auto_patterns": auto_patterns, "note": msg}))
        else:
            print(msg)
        return

    auto_patterns.remove(pattern_id)
    config["self_heal"]["auto_patterns"] = auto_patterns
    _write_config(config)

    _log({
        "timestamp": _now(),
        "event": "canary_demoted",
        "pattern_id": pattern_id,
        "mode": "suggest",
    })
    _emit_alert("info", f"Pattern {pattern_id} demoted to suggest-mode")

    if args.json:
        print(json.dumps({"demoted": pattern_id, "auto_patterns": auto_patterns}))
    else:
        print(f"Demoted '{pattern_id}' to suggest-mode.")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main() -> None:
    parser = argparse.ArgumentParser(description="Healer canary rollout management")
    parser.add_argument("--status", action="store_true", help="Show pattern status (default)")
    parser.add_argument("--promote", metavar="PATTERN_ID", help="Promote pattern to auto-mode")
    parser.add_argument("--demote", metavar="PATTERN_ID", help="Demote pattern to suggest-mode")
    parser.add_argument("--json", dest="json", action="store_true", help="Output JSON")
    args = parser.parse_args()

    if args.promote:
        cmd_promote(args)
    elif args.demote:
        cmd_demote(args)
    else:
        cmd_status(args)


if __name__ == "__main__":
    main()

File: tools/healer_canary.ts
#!/usr/bin/env node
/**
 * healer_canary CLI — TypeScript port of `scripts/healer_canary.py`
 * plus new subcommands.
 *
 * Surface:
 *
 *   healer_canary.ts [--status] [--json]
 *     List every pattern with mode/circuit/eligibility. Same JSON
 *     shape as the Python (consumed by `tools/stark_session_lib.ts`).
 *
 *   healer_canary.ts --promote PATTERN_ID [--json]
 *     Move a pattern from suggest-mode to auto-mode. Fails fast with
 *     explicit unmet criteria if the promotion gate isn't met.
 *
 *   healer_canary.ts --demote PATTERN_ID [--json]
 *     Move a pattern from auto-mode back to suggest-mode.
 *
 *   healer_canary.ts --check [--json]                          (NEW)
 *     Exits 0 if no auto-mode pattern has its circuit open, exits 2
 *     otherwise (and prints the offenders). Designed for oncall paging
 *     — "is the canary still healthy?"
 *
 *   healer_canary.ts --close-circuit PATTERN_ID [--json]       (NEW)
 *     Reset a tripped circuit. Clears `tripped_at` and
 *     `consecutive_failures`, stamps `last_reset_at`, preserves
 *     `ever_tripped` as historical record.
 *
 *   healer_canary.ts --explain PATTERN_ID [--json]             (NEW)
 *     Audit trail for a single pattern: every log entry, current
 *     circuit state, computed stats, mode, eligibility.
 */

import fs from "node:fs";

import {
  cmdCheck,
  cmdCloseCircuit,
  cmdDemote,
  cmdExplain,
  cmdPromote,
  cmdStatus,
} from "./healer_canary_lib.ts";

// ---------------------------------------------------------------------------
// Tiny argv parser
// ---------------------------------------------------------------------------

interface ParsedArgs {
  flags: Map<string, string | true>;
}

function parseArgs(argv: string[]): ParsedArgs {
  const flags = new Map<string, string | true>();
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (!a.startsWith("--")) continue;
    const eq = a.indexOf("=");
    if (eq !== -1) {
      flags.set(a.slice(2, eq), a.slice(eq + 1));
    } else {
      const next = argv[i + 1];
      if (next !== undefined && !next.startsWith("--")) {
        flags.set(a.slice(2), next);
        i++;
      } else {
        flags.set(a.slice(2), true);
      }
    }
  }
  return { flags };
}

function flagString(args: ParsedArgs, name: string): string | undefined {
  const v = args.flags.get(name);
  return typeof v === "string" ? v : undefined;
}

function flagBool(args: ParsedArgs, name: string): boolean {
  return args.flags.has(name);
}

// ---------------------------------------------------------------------------
// Output renderers (text / JSON)
// ---------------------------------------------------------------------------

function pad(s: string, n: number): string {
  return s.length >= n ? s : s + " ".repeat(n - s.length);
}

function renderStatusText(rows: ReturnType<typeof cmdStatus>): string {
  const lines: string[] = [];
  for (const row of rows.patterns) {
    const circuit = row.circuit === "open" ? "OPEN  " : "closed";
    const eligible = row.eligible_for_promotion ? " [eligible]" : "";
    const suggests = `${row.successful_suggests}/5 suggests`;
    lines.push(
      `  ${pad(row.id, 30)}  mode=${pad(row.mode, 7)}  circuit=${circuit}  ${suggests}  rate=${Math.round(row.success_rate * 100)}%${eligible}`,
    );
  }
  return lines.join("\n");
}

function renderCheckText(result: ReturnType<typeof cmdCheck>): string {
  if (result.ok) return `OK: ${result.checked} auto-mode pattern(s), all circuits closed.`;
  const lines: string[] = [
    `FAIL: ${result.tripped_auto_patterns.length} auto-mode pattern(s) have an open circuit:`,
  ];
  for (const id of result.tripped_auto_patterns) lines.push(`  - ${id}`);
  return lines.join("\n");
}

function renderExplainText(result: ReturnType<typeof cmdExplain>): string {
  if (!result.found) return `Pattern '${result.pattern_id}' not found in healer_patterns.json.`;
  const lines: string[] = [];
  lines.push(`Pattern:  ${result.pattern_id}`);
  lines.push(`Mode:     ${result.mode}`);
  const s = result.stats!;
  const circuit = s.circuit_open ? "OPEN" : "closed";
  lines.push(
    `Stats:    total=${s.total_attempts}  suggested=${s.successful_suggests}  applied=${s.applied}  aborts(7d)=${s.aborts_last_7d}  rate=${Math.round(s.success_rate * 100)}%`,
  );
  lines.push(
    `Circuit:  ${circuit}  consecutive_failures=${s.consecutive_failures}  ever_tripped=${s.ever_tripped}${s.tripped_at ? `  tripped_at=${s.tripped_at}` : ""}`,
  );
  if (result.mode === "suggest") {
    const eligible = result.eligible_for_promotion ? "yes" : "no";
    lines.push(`Eligible: ${eligible}`);
    if (result.promotion_blockers && result.promotion_blockers.length > 0) {
      lines.push("Blockers:");
      for (const b of result.promotion_blockers) lines.push(`  - ${b}`);
    }
  }
  lines.push("");
  lines.push(`Log entries (${result.entries.length}):`);
  for (const e of result.entries) {
    const tag = e.status ?? e.event ?? "(unknown)";
    lines.push(`  ${e.timestamp}  ${tag}`);
  }
  return lines.join("\n");
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

const HELP = [
  "usage: healer_canary.ts <one of:>",
  "  --status [--json]",
  "  --promote PATTERN_ID [--json]",
  "  --demote PATTERN_ID [--json]",
  "  --check [--json]                 (oncall: nonzero if auto circuit open)",
  "  --close-circuit PATTERN_ID [--json]",
  "  --explain PATTERN_ID [--json]",
  "",
].join("\n");

function main(argv: string[]): number {
  if (argv.includes("-h") || argv.includes("--help")) {
    process.stderr.write(HELP);
    return 0;
  }
  const args = parseArgs(argv);
  const asJson = flagBool(args, "json");

  const promote = flagString(args, "promote");
  const demote = flagString(args, "demote");
  const closeCircuit = flagString(args, "close-circuit");
  const explain = flagString(args, "explain");
  const check = flagBool(args, "check");

  if (promote) {
    const result = cmdPromote(promote, {});
    if (asJson) {
      process.stdout.write(`${JSON.stringify(result)}\n`);
    } else if (result.ok && result.already_present) {
      process.stdout.write(`Pattern '${promote}' is already in auto_patterns\n`);
    } else if (result.ok) {
      process.stdout.write(`Promoted '${promote}' to auto-mode.\n`);
    } else {
      process.stderr.write(`Cannot promote '${promote}' — criteria not met:\n`);
      for (const r of result.reasons ?? []) process.stderr.write(`  - ${r}\n`);
      return 1;
    }
    return 0;
  }

  if (demote) {
    const result = cmdDemote(demote, {});
    if (asJson) {
      process.stdout.write(`${JSON.stringify(result)}\n`);
    } else if (result.not_present) {
      process.stdout.write(`Pattern '${demote}' is not in auto_patterns\n`);
    } else {
      process.stdout.write(`Demoted '${demote}' to suggest-mode.\n`);
    }
    return 0;
  }

  if (closeCircuit) {
    const result = cmdCloseCircuit(closeCircuit, {});
    if (asJson) {
      process.stdout.write(`${JSON.stringify(result)}\n`);
    } else if (result.no_op) {
      process.stdout.write(`Pattern '${closeCircuit}' circuit was already closed.\n`);
    } else {
      process.stdout.write(`Closed circuit for '${closeCircuit}'.\n`);
    }
    return 0;
  }

  if (explain) {
    const result = cmdExplain(explain, {});
    if (asJson) {
      process.stdout.write(`${JSON.stringify(result)}\n`);
      return result.found ? 0 : 1;
    }
    process.stdout.write(`${renderExplainText(result)}\n`);
    return result.found ? 0 : 1;
  }

  if (check) {
    const result = cmdCheck({});
    if (asJson) {
      process.stdout.write(`${JSON.stringify(result)}\n`);
    } else {
      process.stdout.write(`${renderCheckText(result)}\n`);
    }
    // Non-zero exit when an auto-mode circuit is open — for oncall paging.
    return result.ok ? 0 : 2;
  }

  // Default: status
  const result = cmdStatus({});
  if (asJson) {
    process.stdout.write(`${JSON.stringify({ patterns: result.patterns })}\n`);
  } else {
    process.stdout.write(`${renderStatusText(result)}\n`);
  }
  return 0;
}

function isMain(): boolean {
  try {
    const argv1 = process.argv[1];
    if (!argv1) return false;
    const realArgv = fs.realpathSync(argv1);
    const realModule = fs.realpathSync(new URL(import.meta.url).pathname);
    return realArgv === realModule;
  } catch {
    return false;
  }
}

if (isMain()) {
  try {
    process.exit(main(process.argv.slice(2)));
  } catch (err) {
    process.stderr.write(`healer_canary: ${(err as Error).message}\n`);
    process.exit(1);
  }
}
